### PR TITLE
Add Scheduled VM Snapshot Retention using PingSource Function Example

### PR DESCRIPTION
**Pull request by lamw**
Closes: #627
Signed-off-by: William Lam <wlam@vmware.com>

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [x] Please ensure that you are making a pull request against the **Development** branch
- [ ] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [x] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [x] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commits related to your change
- [x] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [ ] Please include any relevant screenshots and/or output as part of your testing
- [x] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation changes
- [ ] Other (please describe):

### CODE REVIEW

File: examples/knative/powercli/kn-pcli-snapshot-cron/function.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+              value: "true"
+---
+apiVersion: sources.knative.dev/v1beta2
+kind: PingSource
+metadata:
+  name: snapshot-mgmt-ping-source
+spec:
+  schedule: "*/5 * * * *"
+  contentType: "application/json"
+  data: '{"dryRun": true,"retentionConfig": {"sizeGB": "", "days": "", "count": "1"}, "virtualMachines": ["VM-1", "VM-2"]}'
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-snapshot-cron/function.yaml:30]:
```suggestion
  # snapshot management configuration
  data: '{"dryRun": true,"retentionConfig": {"sizeGB": "", "days": "", "count": "1"}, "virtualMachines": ["VM-1", "VM-2"]}'
```

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+   # Extract snapshot retention configuration & VM list from 'data' payload
+   $maxSnapshotCount = ${cloudEventData}.retentionConfig.count
+   $maxSnapshotSize = ${cloudEventData}.retentionConfig.sizeGB
+   $maxSnapshotDays = ${cloudEventData}.retentionConfig.days
+   $vmList = ${cloudEventData}.virtualMachines
+   $dryRun = ${cloudEventData}.dryRun
+
+   # Check to ensure at least one snapshot policy is defined
+   if($maxSnapshotCount -eq "" -and $maxSnapshotSize -eq "" -and $maxSnapshotDays -eq "") {
+      throw "No snapshot retentation policy were defined"
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:84]:
```suggestion
      throw "No snapshot retentation policy defined"
```

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+   # Check to ensure we have a list of VMs
+   if($vmList -eq "" -or $vmList.getType().BaseType.Name -ne "Array") {
+      throw "Invalid or missing input found for VM list"
+   }
+
+   # Retrieve list of VMs to run snapshot management against
+   try {
+      $vms = Get-Vm -NoRecursion -Name $vmList
+   } catch {
+      throw "Failed to retrieve VM list"
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:96]: Will this also include the exception details from `Get-Vm`?

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                     $snapshot | Remove-Snapshot -RunAsync -Confirm:$false
+                  } catch {
+                     throw "Failed to remove snapshot $snapshotName from VM $vmName with error $($Error[0])"
+                  }
+               }
+            }
+         }
+      }
+
+      if($maxSnapshotDays -ne "") {
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:130]: I thought logic is to only perform one if condition, i.e. stop (return) after `maxSnapshotSize` if it is set?

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+         throw "Failed to retrieve snapshots from VM $vmName "
+      }
+
+      if($maxSnapshotSize -ne "") {
+         foreach ($snapshot in $snapshots) {
+            $snapshotName = ${snapshot}.name
+
+            # Get snapshot size (GB) and check against max size
+            if($snapshot.sizeGB -gt $maxSnapshotSize) {
+               Write-Host "$(Get-Date) - `tSnapshot $snapshotName consumes $(${snapshot}.SizeGB) GB and exceeds maximum size (${maxSnapshotSize} GB)"
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:115]: `t < for tab?

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+               if($dryRun -eq $false) {
+                  Write-Host "$(Get-Date) - `tSnapshot removal started for $snapshotName"
+
+                  try {
+                     $snapshot | Remove-Snapshot -RunAsync -Confirm:$false
+                  } catch {
+                     throw "Failed to remove snapshot $snapshotName from VM $vmName with error $($Error[0])"
+                  }
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:124]: This logic is repetitive across the different actions (at least for days/GB) and can be collapsed into a single action before the function returns.

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+               }
+            }
+         }
+      }
+
+      if($maxSnapshotCount -ne "") {
+         if($snapshots.count -gt $maxSnapshotCount) {
+
+            # Get the older snapshot
+            $snapshot = $snapshots | Select-Object -Last 1
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:160]: Shouldn't you remove all snaps (i.e. diff) between max and current.count? Also, "-Last" is sorted by oldest here? For me "-Last" reads like the last (most recent) one :)

File: examples/knative/powercli/kn-pcli-snapshot-cron/test/test-payload.json [hunk 8]
```diff
@@ -0,0 +1,12 @@
+{
+    "dryRun": true,
+    "retentionConfig": {
+        "sizeGB": "",
+        "days": "",
+        "count": "1"
+    },
+	"virtualMachines": [
+		"VM-1",
+        "VM-2"
```
Comment (embano1) [inline examples/knative/powercli/kn-pcli-snapshot-cron/test/test-payload.json:10]: Formatting looks off

**Review by embano1 — CHANGES_REQUESTED** [review]

File: examples/knative/powercli/kn-pcli-snapshot-cron/README.md [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+User          : VSPHERE.LOCAL\Administrator
+Uid           : /VIServer=vsphere.local\administrator@vcsa.primp-industries.local:443/
+Version       : 7.0.2
+Build         : 18455184
+ProductLine   : vpx
+InstanceUuid  : 056a402b-2b3d-4f74-93f7-fa818adff697
+RefCount      : 1
+ExtensionData : VMware.Vim.ServiceInstance
+
+09/29/2021 14:12:11 - Successfully connected to vcsa.primp-industries.local
```
[the remark this answers is not in the document]
Comment (lamw) [reply examples/knative/powercli/kn-pcli-snapshot-cron/README.md:122]: Yea, I noticed this only with my local setup. When deployed to VEBA, its near instantly, haven't been able to figure out why but this is most likely environmental

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 (hunk 5 above)
Comment (lamw) [reply examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:115]: Yup, '`' is PS special escape character :)

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 (hunk 4 above)
Comment (lamw) [reply examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:130]: No, as discussed, user can specify any one of the three params. These (if) checks ensures something was set or else it won't evaluate

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 (hunk 6 above)
Comment (lamw) [reply examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:124]: Good point! Will refactor

File: examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1 (hunk 7 above)
Comment (lamw) [reply examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:160]: hm ... logic is indeed invalid right now. It shouldn't remove all but basically the "diff" as you mentioned. 

```
$diff = $maxSnapshotCount - $snapshots.count
$snapsToRemove = $snapshots | Select-Object -Last $diff
```

This way, you're removing the n-number of oldest but this also brings up a good question on whether users want to delete from newest or oldest within snapshot tree ...

Comment (lamw) [reply examples/knative/powercli/kn-pcli-snapshot-cron/handler.ps1:160]: I think the logic to handle count might be tricky. For sake of simplicity, I'm going to remove this use case

**Review by embano1 — APPROVED** [review]
LGTM

**Pull request merged**